### PR TITLE
Preserve unmodeled spec fields when MultiKueue SyncJob creates remote jobs

### LINKED ISSUE #1234: Avoid use of Consistently in Kubeflow tests
**What would you like to be cleaned**:

The Kubeflow tests (MPI and others) have 5 uses of `Consistently`. We should remove them if they don't provide much value.

**Why is this needed**:

A somewhat busy wait in integration tests can increase the runtime of the tests and usually add meaningful coverage. We should avoid them as much as possible.

### LINKED ISSUE #14837: Preserve unmodeled spec fields when MultiKueue SyncJob creates remote jobs
## Summary

MultiKueue `SyncJob` implementations can drop fields that are not modeled by the vendored API types.

The current pattern copies the remote job specification from `*localJob.Spec.DeepCopy()` and then creates the remote job. Kubernetes removes fields that the typed client does not model during decoding. The created remote job therefore does not contain those fields.

The merge-patch change in #14836 preserves unknown fields when Kueue updates an existing TrainJob. It cannot preserve fields that are already absent when MultiKueue creates the remote job.

## Required changes

Review the built-in MultiKueue job adapters that construct remote jobs from typed `Spec.DeepCopy()` results.

Define and implement a field-preservation approach for remote-job creation. The approach must preserve valid fields that exist in the source job specification but are unknown to Kueue's vendored API version.

Use the existing unstructured-spec-copy pattern, where applicable, as a reference. Keep the existing typed behavior for Kueue-managed fields.

## Rationale

New fields can be accepted and stored by a workload CRD before Kueue upgrades its vendored API dependency. MultiKueue must not silently omit those fields when it creates the corresponding remote job.

## Affected area

- MultiKueue `SyncJob` implementations in built-in job adapters.
- Remote job creation paths that copy typed job specifications.

## Acceptance criteria

- [ ] Remote-job creation preserves source `spec` fields that are not modeled by Kueue's vendored API types.
- [ ] The implementation does not regress Kueue-managed fields or existing MultiKueue synchronization behavior.
- [ ] Tests verify that an unmodeled source `spec` field is present on the created remote job.
- [ ] The implementation documents any adapter-specific limitations that cannot use the common approach.

## References

- TrainJob merge-patch change: https://github.com/kubernetes-sigs/kueue/pull/14836
- Follow-up review: https://github.com/kubernetes-sigs/kueue/pull/14836#pullrequestreview-5043265018
- Requested by: @sohankunkerkar

### PR DESCRIPTION
**Pull request by kavix**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
MultiKueue `SyncJob` implementations previously copied the remote job specification directly from typed Go structs via `*localJob.Spec.DeepCopy()`. This caused fields that were not modeled by Kueue's vendored CRD/API types to be silently dropped when creating the remote job.

This PR introduces:
1. `jobframework.GetUnstructured` and `jobframework.CreateWithPreservedSpec` / `jobframework.NewUnstructuredWithPreservedSpec` in `pkg/controller/jobframework`.
2. A JSON merge patch workflow between the typed local spec and typed transformed destination spec applied over the raw local unstructured spec.
3. Creation of the remote job directly via `*unstructured.Unstructured` to prevent serializer field loss.
4. Updates to all built-in MultiKueue adapters:
   - AppWrapper
   - batch/Job
   - JobSet
   - Kubeflow jobs (JAXJob, PaddleJob, PyTorchJob, TFJob, XGBoostJob)
   - LeaderWorkerSet
   - MPIJob
   - Pod / PodGroup
   - Ray (RayCluster, RayJob, RayService)
   - StatefulSet
   - TrainJob

#### Which issue(s) this PR fixes:
Fixes #14837

#### Special notes for your reviewer:
None

#### Does this PR introduce a user-facing change?
```release-note
Preserve unmodeled spec fields when MultiKueue SyncJob creates remote jobs.
```

## Summary by CodeRabbit

- **New Features**
  - Preserves original workload specification fields when synchronizing resources to remote clusters.
  - Supports Jobs, Pods, JobSets, AppWrappers, StatefulSets, TrainJobs, MPIJobs, Ray jobs, LeaderWorkerSets, and Kubeflow jobs.
  - Retains unknown, nested, and array-based specification fields while applying required updates.

- **Bug Fixes**
  - Prevents unsupported fields from being dropped during remote resource creation.
  - Preserves user runtime patches while removing managed patches from TrainJobs.
  - Improves synchronization error handling.

### CODE REVIEW

**Comment by kavix** [comment]
@tenzen-y can u review this?

File: pkg/controller/jobframework/multikueue_spec.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+			return nil, fmt.Errorf("marshalling typed source spec: %w", err)
+		}
+		destinationJSON, err := json.Marshal(destinationSpec)
+		if err != nil {
+			return nil, fmt.Errorf("marshalling typed destination spec: %w", err)
+		}
+		patch, err := jsonpatch.CreateMergePatch(sourceJSON, destinationJSON)
+		if err != nil {
+			return nil, fmt.Errorf("creating typed spec merge patch: %w", err)
+		}
```
Comment (mszadkow) [inline pkg/controller/jobframework/multikueue_spec.go]: `jsonpatch.CreateMergePatch/MergePatch` replace changed arrays wholesale (RFC 7396) - any array field that differs between typed source/destination wipes unknown raw fields on every element, not just the changed one.

You can use this prompt to generate the unit test exposing it:
```
In pkg/controller/jobframework/multikueue_spec.go, NewUnstructuredWithPreservedSpec computes a merge patch between the typed source/destination specs and applies it to the raw spec. 
JSON merge patch replaces changed arrays wholesale, so if only one container's image changes, the whole containers array gets replaced by the typed version — wiping unknown fields on untouched sibling elements. 
Write a Go test in multikueue_spec_test.go using a batchv1.Job with two containers ("main", "sidecar"): destination changes only main's image; rawSource has an extra "futureField" on sidecar (untouched). 
Assert containers[0].image == "app:v2" and containers[1].futureField == "keep-me" survives — it currently doesn't.

```

Comment (kavix) [reply pkg/controller/jobframework/multikueue_spec.go]: Thanks for pointing this out and for the test case @mszadkow!

I have replaced the RFC 7396 merge patch with a 3-way structural merge in `NewUnstructuredWithPreservedSpec`. It matches list items by common merge keys (`name`, `manager`, `key`, `containerPort`, etc.) so that touched elements merge cleanly while untouched sibling elements and unmodeled fields are preserved. I have also added the suggested unit test to verify this.
